### ISSUE TITLE
Special name for the current package: `@import("self")`

### ISSUE THREAD

**Issue by xackus** [ziglang/zig#6279]
Proposal: `"self"` becomes a special package name (like `"root"` currently is), referring to the current package.

This allows using `@import("self")` in any file, without specifying a relative path (`"../../packagel.zig"`).

> For now, the previous way to refer to the library is the preferred way. The only reason std has itself in its own package map is for this check: 
> 
> https://github.com/ziglang/zig/blob/a496f94be92581cdf97a1bd60c4dea0369e26395/lib/std/os.zig#L42
> 
> I'd like to keep it referring to std.zig because that is what user made packages will have to do and it's nice to keep the std lib code to match the same thing as an example.
> 
> If you like we can do a proposal for a special name to be used for a package to refer to itself. For example we already have `@import("root")` for referring to the application's root source file, but we could also introduce `@import("self")` which would refer to its own package. Then for the std lib we would no longer put "std" in the package table; we would rely on the "self" feature, which would be available to all packages. Are you interested in making that proposal?
> 
_Originally posted by @andrewrk in https://github.com/ziglang/zig/pull/6278#discussion_r484555160_

**Comment by andrewrk** [comment]
Accepting now as I think this will be uncontroversial and I don't really see any downsides.

**Comment by MasterQ32** [comment]
May i add a follow-up proposal that could be included here:

As `@import` is a built-in, it might allow both strings (*always* user packages) and enum literals (`.root`, `.self`, `.std`) to be imported:
```zig
const network = @import("network");
const std = @import(.std);
```

This would make clear that these packages are *magic* and provided by the compiler whereas every string is to be given on the command line or in `build.zig`. It would also free those magic package names

**Comment by xackus** [comment]
@MasterQ32: I don't think this can be done with `std` because the types used in builtins reside in `std.builtin`.

**Comment by MasterQ32** [comment]
@xackus Builtins (as in functions) can accept any type. In this case: A *enum literal*, not value from a explicit. This means it should work... But some of the compiler/language devs should think about it as well

**Comment by daurnimator** [comment]
@xackus / @MasterQ32  even a normal zig function can do that:
```zig
fn somefunc(a: anytype) type {
  switch(@typeInfo(@TypeOf(a))) {
    .Pointer => .....,
    .EnumLiteral => ....,
    else => @compileError("invalid type to import"),
  }
}
```

**Comment by andrewrk** [comment]
@MasterQ32 here's a good place for your follow-up proposal: https://github.com/ziglang/zig/issues/2206

**Comment by nektro** [comment]
maybe a builtin similar to `@This()` would be a better approach

**Comment by zzyxyzz** [comment]
> maybe a builtin similar to `@This()` would be a better approach

Isn't a top-level `@This()` already equivalent to importing "self"?

**Comment by leecannon** [comment]
@zzyxyzz this proposal is about a way of importing the root of the current package not the current file.

This does highlight that importing "self" clashes with an already well used idiom of `const Self = @This()`;

**Comment by nektro** [comment]
no one would use the name `self` or `Self` for the variable. they'd use the name of their package.

**Comment by zzyxyzz** [comment]
> this proposal is about a way of importing the root of the current package not the current file.

Right. Sorry.

Though I wasn't aware that Zig had the concept of a package root. Isn't it anonymous structs all the way down?

**Comment by presentfactory** [comment]
Is this still planned? I would like to use this too because currently the only way to import a package within itself without issues is to use a relative path as mentioned previously and that is undesirable especially when you get into deeper folders.

Another option if the self thing has issues could be to allow letting you specify the package you're in by its name, for example if the Package `foo` refers to `foo.zig` then allow `foo.zig` or any file imported by it to do `@import("foo")` (as currently this fails and you have to do `@import("foo.zig")`).

**Comment by billzez** [comment]
+1

Title should probably be updated to "current module" now.

**Comment by deanveloper** [comment]
> I think that `self` might not be the best name

A few alternate names that could be used are `mod`, `module`, and `module-root`. I'm not sure if the accepted label was removed _because_ of the naming issue or if the removal was for different reasons, but I'd love for this functionality to be in Zig as someone who enjoys creating libraries. The special exception so that `std` can refer to itself with `@import("std")` was added to address this concern, and it'd be really nice if third-party libraries could do this too 😄

**Comment by Lking03x** [comment]
Upon encountering the inconvenience, I first thought that `@import("root/src_sub_folder/file.zig")` would be the way to go however that seems a wrong idea giving the actual semantic of "root" module.

What is needed is the ability to import files relative to the package main source directory **as seen in the filesystem hierarchy**. As I understand from the discussion "self" will just act like "root" and will rely on the `pub` / "non pub" of declarations access contains of other files (in other sub directories). I thus think a `const self = @import("self")` won't solve the problem regardless of the name. Same for a builtin similar to `@This()`.

I propose:
1. `@import("package_name/sub_dir_of_src/file.zig")`
    - "package_name" is less likely to collide with an existing module name as it's the actual name of package current get from `build.zig.zon`. 
    - "sub_dir_of_src" is relative to src and follow the filesystem just like relative import does.
    for example `@import("ziggraph/scanner/utils.zig")` or `@import("ziggraph/ui/screen/test_screen.zig")`
    This is similar to how the dart language and others resolve their import.

2. The approach of using the actual package name may lead to several problem.
    - Some package use special character in their name ("raylib-zig")
    - And it can becomes hard to differentiate which module is the current package
    **Instead of** "package_name" we may have a generic **"package"** (which is more explicit than having a symbol like "/"). A module named "package" is also less likely to cause conflict.

    for example `@import("package/scanner/utils.zig")` or `@import("package/ui/screen/test_screen.zig")` but also `@import("package/main.zig")` or `@import("package/root.zig")`.

    Either `const package = @import("package")` itself won't be legal or if legal will point to the root file of the current package. In that case ` @import("package")` == ` @import("package/root.zig")` for libraries.
    `const imported = @import("package/sub_path/importable_file.ext")` will always be legal. "ext" maybe `zig` or `zon`.

The main point is that we need to be able to import any importable file with a path like syntax relative to the current package directory

**Comment by Lking03x** [comment]
Upon encountering the inconvenience, I first thought that `@import("root/src_sub_folder/file.zig")` would be the way to go however that seems a wrong idea giving the actual semantic of "root" module.

What is needed is the ability to import files relative to the package main source directory **as seen in the filesystem hierarchy**. As I understand from the discussion "self" will just act like "root" and will rely on the `pub` / "non pub" of declarations access contains of other files (in other sub directories). I thus think a `const self = @import("self")` won't solve the problem regardless of the name. Same for a builtin similar to `@This()`.

I propose:
1. `@import("package_name/sub_dir_of_src/file.ext")`
    - "package_name" is less likely to collide with an existing module name as it's the actual name of package current get from `build.zig.zon`. 
    - "sub_dir_of_src" is relative to src and follow the filesystem just like relative import does.
    - "ext" maybe "zig" or "zon".
    for example `@import("ziggraph/scanner/utils.zig")` or `@import("ziggraph/ui/screen/test_screen.zig")`
    This is similar to how the dart language and others resolve their import.

2. The approach of using the actual package name may lead to several problem.
    - Some package use special character in their name ("raylib-zig")
    - And it can becomes hard to differentiate which module is the current package
    **Instead of** "package_name" we may have a generic **"package"** (which is more explicit than having a symbol like "/"). A module named "package" is also less likely to cause conflict.

    for example `@import("package/scanner/utils.zig")` or `@import("package/ui/screen/test_screen.zig")` but also `@import("package/main.zig")` or `@import("package/root.zig")`.

    **Either**
    - `const package = @import("package")` itself won't be legal. In that case `@import("package")` will only refers to the current package folder, the first folder containing "build.zig" and "build.zig.zon" will walking up the filesystem hierarchy from the caller path. And `@import("package/ui/screen/test_screen.zig")` becomes `@import("package/src/ui/screen/test_screen.zig")`.
    - or if legal will point to the root file of the current package. In that case ` @import("package")` == ` @import("package/root.zig")` for libraries, assuming "root.zig" is in "package_folder/src". I'm not sure if it's the best way, some packages have ".zig" files in "/lib" sub-directory.

The main point is that we need to be able to import any importable file with a path like syntax relative to the current package directory
